Uses local RPM build for "dev" and "staging" scenarios

conorsch opened the pull request:
## Status

Ready for review 

## Description of Changes

Fixes #538 

Changes proposed in this pull request: 

* Using the "make dev" environment now uses RPMs for install/uninstall
* Docker is now required in the dev env!

Hopefully this will make reasoning about problems like #505 a lot more accessible to developers. 

## Testing
Requesting testing for multiple scenarios, since early review showed some surprises. 

### Dev scenario
There's a catch here—you must run `make clone` _twice_ in order to get started, because the first clone will only pull in the new docker logic, not run it. 

```
make clone
make clone # [sic]
make clean # ensure everything removed cleanly
make dev
make test
```

Mess around with the clean/clone/dev/test combinations and see if you can get anything to break. 

### Staging scenario
Now let's check that "staging" scenario continues to work well. Notably the locally built RPM will be installed in staging. 


```
make clone
make clone # [sic, just in case you didn't already run it twice for dev]]
make clean # ensure everything removed cleanly
make staging
```

Interact, confirm staging-specific functionality like shutdown-on-lid close is working, then run `make clean` to clean up and confirm no errors.

#### Prod scenario
Since the prod flow uses a separate setup process (https://workstation.securedrop.org/en/stable/admin/install.html#download-and-install-securedrop-workstation), there shouldn't be any conflicts with the logic presented here. If you can think of some, please comment or directly edit this test plan!

## Checklist

### If you have made code changes

- [ ] Linter (`make flake8`) passes in the development environment (this box may
      be left unchecked, as `flake8` also runs in CI)

### If you have made changes to the provisioning logic

- [ ] All tests (`make test`) pass in `dom0` of a Qubes install

- [ ] This PR adds/removes files, and includes required updates to the packaging
      logic in `MANIFEST.in` and `rpm-build/SPECS/securedrop-workstation-dom0-config.spec`

conorsch commented:
Marking as ready-for-review. Rebased these changes and tested locally, and was pleased with the results, so it's ready for more eyes.

conorsch commented on dom0/sd-dom0-files.sls:
```diff
@@ -198,21 +198,13 @@ dom0-securedrop-launcher-desktop-shortcut:
     - mode: 755
 
 {% import_json "sd/config.json" as d %}
-{% if d.environment == "dev" %}
```
That's true! Closely related to your comment (https://github.com/freedomofpress/securedrop-workstation/pull/587/files#r458880338) about version logic. The behavior you describe is the least surprising, in my mind. New packages will _only_ be installed from the test rpm server when a newer semver is available there, presumably during QA. Running the GUI updater should never be a no-op, in my mind, so allowing package versions to supersede makes the most sense. We could entirely disable the repos for the dev env, but that sounds like a step backwards—curious if you see it differently!

conorsch commented on scripts/clone-to-dom0:
```diff
@@ -20,12 +20,19 @@ dev_dir="${SECUREDROP_DEV_DIR:-/home/user/securedrop-workstation}"
+# in the subsequent tarball, which is fetched to dom0.
+function build-dom0-rpm() {
+    printf "Building RPM on %s ...\n" "${dev_vm}"
+    qvm-run -q "$dev_vm" "make -C $dev_dir dom0-rpm"
```
I'm not aware of any RPM tooling like `dch` for programmatically bumping the version. Ideally it'd be something like `<current_version><alpha1><git_short_hash>`, but absent tooling, simply defaulting to `<current_version>` makes the most sense to me.

emkll reviewed (dismissed):
Went through this again, from a development scenario perspective, changes look good and are a significant improvement for developers and brings more consistency across environments.

However it looks like there might be some implications for prod/staging scenarios (see inline). It might make sense to only invoke `prep-salt` exclusively for the dev scenario (instead of conditionals).

There's also a potential versioning issue: if the version of the local RPM is lower than the one served by the rpm repo, changes to the salt files may be overwritten by the ones on the RPM servers. We can autoupdate but we can also just provide a docs update to unblock, since this will only affect developers.

Finally, tests rely on the `create_tarball` function https://github.com/freedomofpress/securedrop-workstation/pull/587/files#diff-3c3ec2225864916f85515009ebeeb428L24 . Packaging the tests and reducing/eliminating the requirement of a full clone of the repository in dom0 may help reduce developer error (changes to salt files in the dom0 securedrop-workstation folder will effectively be useless). I don't think this should be immediately addressed as part of this PR, but should probably be tracked in a  follow-up issue

UPDATE: we will need to rebase and update references to the installer (`securedrop-admin` -> `sdw-admin` since  https://github.com/freedomofpress/securedrop-workstation/pull/596 has been merged)

eloquence commented:
We should clarify the README as part of this PR, @conorsch if you want I can add a commit to take a stab at it once other comments are addressed.

eloquence commented:
(Needs rebase after #596)

conorsch commented on scripts/prep-salt:
```diff
@@ -1,39 +1,30 @@
+    exit 1
 fi
 
+echo "Deploying Salt config..."
```
> At second glance, it looks like we may not even need prep-salt for staging and production scenarios, so we can simply remove them and rename prep-salt to something like prep-dev

Great suggestion! I've appended a commit doing just that. :+1:

conorsch commented on scripts/clean-salt:
```diff
@@ -1,27 +0,0 @@
-#!/bin/bash
```
>  especially if we are supporting the make $appvm targets for quick testing

Separate discussion, I suppose, but those targets don't appear well maintained to me, and I think we should remove them. Resisting doing so as part of this PR in an effort to keep the diff manageable.

conorsch commented on Makefile:
```diff
@@ -110,12 +108,9 @@ remove-sd-log: assert-dom0 ## Destroys SD logging VM
-	sudo dnf -y -q remove securedrop-workstation-dom0-config 2>/dev/null || true
-	$(MAKE) clean-salt
+# Use the local script path, since system PATH location will be absent
+# if clean has already been run.
```
Didn't make any changes here. The prep-dev (previously "prep-salt") target does itself call validate, so I think we're covered. On multiple runs of the dev env, I couldn't get this part to error, but perhaps I'm missing something.

conorsch commented:
Rebased to resolve conflicts, and implement a few requested changes. Ready for review!

@eloquence To your point about the README, thank you, I'd welcome fresh eyes on docs updates here.

conorsch commented:
Downgraded to draft, to block merge until we've got the docs updated. Still ready for functional review.

eloquence commented on scripts/prep-dev:
```diff
@@ -0,0 +1,30 @@
+sudo dnf install -y "$latest_rpm"
+
+# Always copy secrets, these won't be configured as part
+# of the RPM installation.
```
Maybe add an `echo` for these operations as well, to make it clear that these files have been copied to a new destination?

conorsch replied:
Great suggestion, added!

eloquence commented on scripts/sdw-admin.py:
```diff
@@ -39,6 +39,20 @@ def parse_args():
+        default=False,
+        required=False,
+        action="store_true",
+        help="During uninstall action, leave TemplateVM RPM packag installed in dom0",
```
Typo: `packag`

conorsch replied:
Thanks, amended and fixed

eloquence commented:
RPM built and copied successfully, but on my first `make dev` run, I am getting the following errors:

```
Set up dom0 config files, including RPC policies, and create VMs
[ERROR   ] Unable to cache file 'salt://securedrop-handle-upgrade' from saltenv 'base'.
[ERROR   ] {'cache_error': True, 'pid': 0, 'retcode': 1, 'stderr': '', 'stdout': ''}
[ERROR   ] Unable to cache file 'salt://securedrop-handle-upgrade' from saltenv 'base'.
[ERROR   ] {'cache_error': True, 'pid': 0, 'retcode': 1, 'stderr': '', 'stdout': ''}

....
----------
          ID: run-prep-upgrade-scripts
    Function: cmd.script
        Name: salt://securedrop-handle-upgrade
      Result: False
     Comment: Unable to cache script salt://securedrop-handle-upgrade from saltenv 'base'
     Started: 17:20:09.559197
    Duration: 6.251 ms
     Changes:   
              ----------
```

and then cascading failures from there, not sure yet if related to the changes in this PR or not.

conorsch commented:
Hmm, those are new errors to me. Can you try a `make clean` and then `make dev`? Perhaps we'll have to _require_ a clean in order to transition to the RPM setup.

eloquence commented:
Doing that now. Noting that `make clean` output starts with _installing_ the package. Is that a precaution, or unintentional?

```
make clean
Deploying Salt config...
Installing RPM at /home/erik/securedrop-workstation/rpm-build/RPMS/noarch/securedrop-workstation-dom0-config-0.4.1-1.fc25.noarch.rpm ...
Qubes OS Repository for Dom0                                                                                                                                   72 MB/s |  74 kB     00:00    
Dependencies resolved.
==============================================================================================================================================================================================
 Package                                                        Arch                               Version                                     Repository                                Size
==============================================================================================================================================================================================
Installing:
 securedrop-workstation-dom0-config                             noarch                             0.4.1-1.fc25                                @commandline                             102 k

Transaction Summary
==============================================================================================================================================================================================
Install  1 Package

Total size: 102 k
Installed size: 335 k
Downloading Packages:
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Installing  : securedrop-workstation-dom0-config-0.4.1-1.fc25.noarch                                                                                                                    1/1 
  Verifying   : securedrop-workstation-dom0-config-0.4.1-1.fc25.noarch                                                                                                                    1/1 

Installed:
  securedrop-workstation-dom0-config.noarch 0.4.1-1.fc25                                                                                                                                      

Complete!
./scripts/sdw-admin.py --uninstall --keep-template-rpm --force
```

eloquence commented:
Unfortunately getting [these errors](https://github.com/freedomofpress/securedrop-workstation/pull/587#issuecomment-670256824) reliably with `make dev` based on this branch, even after repeated `make clean`. Will do a quick `make dev` run on `main` to see if it's related to the changes in this PR or not.

eloquence commented:
`make dev` on `main` is running without these errors (maybe it'll hit others, still running). After that finishes, will see if clearing Salt cache manually after `make clean` makes a difference.

eloquence commented:
Unfortunately even after removing the Salt cache (`sudo rm -rf /var/cache/salt`) and resyncing (`sudo qubesctl saltutil.sync_all`) I'm getting the [same errors](https://github.com/freedomofpress/securedrop-workstation/pull/587#issuecomment-670256824) reproducibly when running `make dev` from this branch following the test plan.

eloquence commented:
@emkll asked about my contents of `/srv/salt` after `make clean`; I'm noticing a dead symlink to `sd-workstation.top` remained in the directory. After re-running `make clean` from the `main` branch of the codebase, that symlink does get removed correctly.

Unfortunately, `make dev` still fails with the previously reported error even after doing so:

```
[ERROR   ] Unable to cache file 'salt://securedrop-handle-upgrade' from saltenv 'base'.
```

<details>
<summary>Contents of <code>/srv/salt</code> after running <code>make clean</code> from this PR's branch</summary>

```
.
├── _grains
│   ├── boot_mode.py
│   ├── boot_mode.pyc
│   ├── boot_mode.pyo
│   ├── pci_devs.py
│   ├── pci_devs.pyc
│   ├── pci_devs.pyo
│   ├── redefined_dom0_grains.py
│   ├── redefined_dom0_grains.pyc
│   ├── redefined_dom0_grains.pyo
│   ├── whonix.py
│   ├── whonix.pyc
│   └── whonix.pyo
├── _modules
│   ├── debug.py
│   ├── debug.pyc
│   ├── debug.pyo
│   ├── ext_module_qvm.py
│   ├── ext_module_qvm.pyc
│   ├── ext_module_qvm.pyo
│   ├── localemod.py
│   ├── localemod.pyc
│   ├── localemod.pyo
│   ├── module_utils.py
│   ├── module_utils.pyc
│   ├── module_utils.pyo
│   ├── qubes_dom0_update.py
│   ├── qubes_dom0_update.pyc
│   ├── qubes_dom0_update.pyo
│   ├── qubes.py
│   ├── qubes.pyc
│   ├── qubes.pyo
│   ├── topd.py
│   ├── topd.pyc
│   └── topd.pyo
├── _pillar
│   ├── qvm_prefs.py
│   ├── qvm_prefs.pyc
│   └── qvm_prefs.pyo
├── qubes
│   ├── directories.sls
│   ├── directories.top
│   ├── files
│   │   ├── locale.sls
│   │   ├── pillar.sls
│   │   └── top.sls
│   ├── LICENSE
│   ├── README.rst
│   ├── user-dirs.sls
│   └── user-dirs.top
├── _states
│   ├── debug.py
│   ├── debug.pyc
│   ├── debug.pyo
│   ├── ext_state_qvm.py
│   ├── ext_state_qvm.pyc
│   ├── ext_state_qvm.pyo
│   ├── status.py
│   ├── status.pyc
│   └── status.pyo
├── topd
│   ├── init.conf
│   ├── init.sls
│   ├── init.top
│   ├── LICENSE
│   └── README.rst
├── top.jinja
├── _tops
│   └── base
│       ├── sd-workstation.top -> /srv/salt/sd-workstation.top
│       └── topd.top -> /srv/salt/topd/init.top
├── top.sls
├── top.sls.old
└── _utils
    ├── fileinfo.py
    ├── fileinfo.pyc
    ├── fileinfo.pyo
    ├── __init__.py
    ├── __init__.pyc
    ├── __init__.pyo
    ├── matcher.py
    ├── matcher.pyc
    ├── matcher.pyo
    ├── nulltype.py
    ├── nulltype.pyc
    ├── nulltype.pyo
    ├── pathinfo.py
    ├── pathinfo.pyc
    ├── pathinfo.pyo
    ├── pathutils.py
    ├── pathutils.pyc
    ├── pathutils.pyo
    ├── qubes_utils.py
    ├── qubes_utils.pyc
    ├── qubes_utils.pyo
    ├── toputils.py
    ├── toputils.pyc
    └── toputils.pyo
```
</details>

<details>
<summary>Contents of <code>/srv/salt</code> after running <code>make clean</code> from the <code>main</code> branch</summary>

```
.
├── dom0-xfce-desktop-file.j2
├── _grains
│   ├── boot_mode.py
│   ├── boot_mode.pyc
│   ├── boot_mode.pyo
│   ├── pci_devs.py
│   ├── pci_devs.pyc
│   ├── pci_devs.pyo
│   ├── redefined_dom0_grains.py
│   ├── redefined_dom0_grains.pyc
│   ├── redefined_dom0_grains.pyo
│   ├── whonix.py
│   ├── whonix.pyc
│   └── whonix.pyo
├── _modules
│   ├── debug.py
│   ├── debug.pyc
│   ├── debug.pyo
│   ├── ext_module_qvm.py
│   ├── ext_module_qvm.pyc
│   ├── ext_module_qvm.pyo
│   ├── localemod.py
│   ├── localemod.pyc
│   ├── localemod.pyo
│   ├── module_utils.py
│   ├── module_utils.pyc
│   ├── module_utils.pyo
│   ├── qubes_dom0_update.py
│   ├── qubes_dom0_update.pyc
│   ├── qubes_dom0_update.pyo
│   ├── qubes.py
│   ├── qubes.pyc
│   ├── qubes.pyo
│   ├── topd.py
│   ├── topd.pyc
│   └── topd.pyo
├── _pillar
│   ├── qvm_prefs.py
│   ├── qvm_prefs.pyc
│   └── qvm_prefs.pyo
├── qubes
│   ├── directories.sls
│   ├── directories.top
│   ├── files
│   │   ├── locale.sls
│   │   ├── pillar.sls
│   │   └── top.sls
│   ├── LICENSE
│   ├── README.rst
│   ├── user-dirs.sls
│   └── user-dirs.top
├── _states
│   ├── debug.py
│   ├── debug.pyc
│   ├── debug.pyo
│   ├── ext_state_qvm.py
│   ├── ext_state_qvm.pyc
│   ├── ext_state_qvm.pyo
│   ├── status.py
│   ├── status.pyc
│   └── status.pyo
├── topd
│   ├── init.conf
│   ├── init.sls
│   ├── init.top
│   ├── LICENSE
│   └── README.rst
├── top.jinja
├── _tops
│   └── base
│       └── topd.top -> /srv/salt/topd/init.top
├── top.sls
├── top.sls.old
└── _utils
    ├── fileinfo.py
    ├── fileinfo.pyc
    ├── fileinfo.pyo
    ├── __init__.py
    ├── __init__.pyc
    ├── __init__.pyo
    ├── matcher.py
    ├── matcher.pyc
    ├── matcher.pyo
    ├── nulltype.py
    ├── nulltype.pyc
    ├── nulltype.pyo
    ├── pathinfo.py
    ├── pathinfo.pyc
    ├── pathinfo.pyo
    ├── pathutils.py
    ├── pathutils.pyc
    ├── pathutils.pyo
    ├── qubes_utils.py
    ├── qubes_utils.pyc
    ├── qubes_utils.pyo
    ├── toputils.py
    ├── toputils.pyc
    └── toputils.pyo
```
</details>

conorsch commented:
@eloquence Maybe you have older rpms lying around from previous testing, and that's affecting the logic. Try this:

```
$ find -type f -iname '*.rpm'
./rpm-build/RPMS/noarch/securedrop-workstation-dom0-config-0.4.0-1.fc25.noarch.rpm
./rpm-build/SRPMS/securedrop-workstation-dom0-config-0.4.0-1.fc25.src.rpm
```

In the error output you helpfully showed, I see version `0.4.1`, which I don't believe ever got merged. If you see hits for 0.4.1 in your local git repo, then a `git clean -idx` should be helpful in straightening things out—just make sure to preserve `config.json` & `sd-journalist.sec`!

eloquence commented:
Yep, looks like that was it -- old RPM with version bump still around that I was using to test #596! Thanks for spotting that.  Will prepare a small docs commit now, after verifying functionality.

eloquence commented on scripts/prep-dev:
```diff
@@ -0,0 +1,31 @@
+dom0_dev_dir="$HOME/securedrop-workstation"
+
+function find_latest_rpm() {
+    find "${dom0_dev_dir}/rpm-build/RPMS/" -type f -iname '*.rpm' | sort -n | tail -n 1
```
Could it make sense to prioritize the most recently built file (by timestamp) instead of going by version number? That way, if a `make clone` run was successful, you'd always know that you'd get exactly the version that was just built. I can see the kind of issue I experienced during review (with an old test RPM taking precedence over a more recent version) coming up again in practice, as we may sometimes want to bump the RPM version temporarily as part of a test plan as we did with #596.

conorsch replied:
> Could it make sense to prioritize the most recently built file (by timestamp) instead of going by version number?

Great suggestion! Added.

eloquence commented:
Initial results of `make test` are looking good. Getting a test failure on `test_log_dirs_properly_named`, but that's #583 (a directory named `host` exist but shouldn't). Otherwise no test failures.

eloquence commented:
First attempt to run `make staging` after a `make clean` on this branch failed as follows:

```
$ make staging
./scripts/configure-environment --env staging
WARNING: Config environment 'staging' was requested.
WARNING: The staging env should only be used on TEST HARDWARE.
WARNING: If you are on a primary laptop for work/production use, 
WARNING: please update your config.json with environment=prod.
WARNING: Are you sure you wish to continue? [y/N] y
Updated config environment to 'staging'...
make validate
make[1]: Entering directory '/home/erik/securedrop-workstation'
make[1]: Leaving directory '/home/erik/securedrop-workstation'
./scripts/provision-all
Configure Fedora-based system VMs
[ERROR   ] Template was specified incorrectly: False
local:
    Data failed to compile:
----------
    No matching sls found for 'sd-sys-vms' in env 'base'
DOM0 configuration failed, not continuing
Makefile:31: recipe for target 'staging' failed
make: *** [staging] Error 1
```

Comparing results with `main` branch now.

conorsch commented:
@eloquence Perhaps we should remove the `make staging` target entirely—the docs still recommend mimicking the prod workflow by fetching the package and ferrying to dom0: https://github.com/freedomofpress/securedrop-workstation/tree/fc44abe1a3441ea84d4856423f18e2d20402f45b#staging-environment

eloquence commented:
`staging` target on `main` appears to be working. Perhaps the breakage is related to the `prep-salt`/`prep-dev` changes in this PR.

> Perhaps we should remove the make staging target entirely—the docs still recommend mimicking the prod workflow by fetching the package and ferrying to dom0:

Right now the README offers both options (under the heading <s>"Choose your own adventure"</s> "Choose your installation method").

Instead of removing the `staging` target, would there be drawbacks to having it bootstrap from the locally built RPM, like `make dev` does? That way you could still quickly test changes that are only visible in `staging` (e.g., to Qubes power management) locally.

eloquence commented:
First set of docs updates appended in 58cedea031b47f2c7ed150d61fbe687ddd7bb484 - hope those edits make sense.

eloquence commented:
I'm getting the [same error](https://github.com/freedomofpress/securedrop-workstation/pull/587#issuecomment-673761126) as for `make staging` with `make prod` on this branch. As with `staging`, I think there's the question of what the future of that Makefile target should be, since it is not recommended to create an actual production install by this route.

eloquence commented:
Per our discussion at standup, I've added a commit 9dc287a14381a813a306c0a9cd346dfbe30839e4 that updates the `staging` target to use a locally built RPM, just like the `dev` target. The intended remaining use of this target is during the development of RCs, and to test differences in behavior that are specific to `staging` (e.g., the power management settings) during development.

That commit also removes the `prod` target in its entirety, since it serves no obvious purpose at this point. I've tested `make staging` locally and it seems to work, and I've updated the README, but I haven't updated the test plan yet. @conorsch If this commit looks good to you, I can flesh out the test plan a bit.

While nightly RPMs are currently "temporarily" disabled, I would propose to block merge of this PR until https://github.com/freedomofpress/securedrop-debian-packaging/issues/186 is resolved, as the current CircleCI configuration still suggests that we may want to re-enable nightly RPM builds at a later date, which would be unsafe to do in combination with this change.

emkll commented on Makefile:
```diff
@@ -11,80 +11,65 @@ all: assert-dom0
 	sudo qubesctl --show-output --skip-dom0 --targets sd-workstation-buster-template state.highstate
 
-sd-proxy: prep-salt ## Provisions SD Proxy VM
+sd-proxy: prep-dev ## Provisions SD Proxy VM
```
The changes to the dev environment may have repercussions on these individual makefile targets :

`prep-salt` makefile target would copy the local salt file in `/srv/salt/` whereas this updated `prep-dev` target will install the dom0 rpm. This means that any changes to the local files in the `securedrop-workstation` folder in dom0 will not be used. 

If this is the case, adding a note to this effect in the dev docs could be helpful

conorsch replied:
> If this is the case, adding a note to this effect in the dev docs could be helpful

That's definitely true. Editing the files in e.g. /srv/salt/ still works fine, but I wouldn't recommend either, given how easy it is to lose changes that way. Will clarify in the docs!

eloquence commented on README.md:
````diff
@@ -313,7 +313,7 @@ make clone
 ```
 
-In the future, we plan on shipping a *SecureDrop Workstation* installer package as an RPM package in `dom0` to automatically update the salt provisioning logic.
+The `make clone` command will build a new version of the RPM package that contains the provisioning logic and copy it to `dom0`.
````
This is stated in the setup instructions for `sd-dev`:
>  You must install Docker in that VM in order to build a development environment using the standard workflow.

But I agree a parenthetical here won't hurt as well, will add a commit to that effect.

eloquence replied:
A bit more detail added in 902fc184170ff978280e9ba6a041353eec93d868

conorsch commented on README.md:
```diff
@@ -192,9 +192,9 @@ In the Qubes Menu, navigate to `System Tools` and click on `Qubes Update`. Click
 You can install the staging environment in two ways:
 
-- If you have an up-to-date clone of this repo with a valid configuration in `dom0`, you can use the `make staging` target to provision a staging environment. Prior to provisioning, `make staging` will set your `config.json` environment to `staging`. As part of the provisioning, your package repository configuration will be updated to use the latest test release of the RPM package, and the latest ...
+- If you have an up-to-date clone of this repo with a valid configuration in `dom0`, you can use the `make staging` target to provision a staging environment. Prior to provisioning, `make staging` will set your `config.json` environment to `staging`. As part of the provisioning, a locally built RPM will be installed in dom0, and your package repository configuration will be updated to use the lat ...
```
Fair point. Rephrased as:

> As part of the provisioning, a locally built RPM will be installed in dom0. The dom0 package repository configuration will be updated to install future test-only versions of the RPM package from the https://yum-test.securedrop.org repository, and Workstation VMs will receive the latest nightlies of the Debian packages (same as `make dev`).

conorsch commented:
Added commits addressing requests for further docs clarifications. Updated the test plan to include staging references. Marking ready for final review.

zenmonkeykstop commented:
Running through test scenarios on a fresh Qubes install (no previous dev env), `make clean` fails with following errors due to `/srv/salt/sd/config.json` not being present:

```
local:
    Data failed to compile:
----------
    Rendering SLS 'base:sd-clean-all' failed: Jinja error: sd/config.json
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/utils/templates.py", line 410, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/site-packages/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 4, in top-level template code
  File "/usr/lib/python2.7/site-packages/salt/utils/jinja.py", line 145, in get_source
    raise TemplateNotFound(template)
TemplateNotFound: sd/config.json

; line 4

---
# -*- coding: utf-8 -*-
# vim: set syntax=yaml ts=2 sw=2 sts=2 et :

{% import_json "sd/config.json" as d %}    <======================

set-fedora-as-default-dispvm:
  cmd.run:
    - name: qvm-check fedora-31-dvm && qubes-prefs default_dispvm fedora-31-dvm || qubes-prefs default_dispvm ''

[...]
---
DOM0 configuration failed, not continuing
Traceback (most recent call last):
  File "./scripts/sdw-admin.py", line 128, in perform_uninstall
    subprocess.check_call(["sudo", "qubesctl", "state.sls", "sd-clean-all"])
  File "/usr/lib64/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['sudo', 'qubesctl', 'state.sls', 'sd-clean-all']' returned non-zero exit status 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./scripts/sdw-admin.py", line 187, in <module>
    main()
  File "./scripts/sdw-admin.py", line 177, in main
    perform_uninstall(keep_template_rpm=args.keep_template_rpm)
  File "./scripts/sdw-admin.py", line 146, in perform_uninstall
    raise SDWAdminException("Error during uninstall")
__main__.SDWAdminException: Error during uninstall
Makefile:100: recipe for target 'clean' failed
make: *** [clean] Error 1
```

zenmonkeykstop commented:
copying `config.json` into place allows `make clean` to finish, but `make dev` fails with the following error:
```
[...]
Installed:
  securedrop-workstation-dom0-config.noarch 0.4.0-1.fc25                                      

Complete!
Copying config secrets into place...
'config.json' -> '/usr/share/securedrop-workstation-dom0-config/config.json'
'sd-journalist.sec' -> '/usr/share/securedrop-workstation-dom0-config/sd-journalist.sec'
WARNING: v2 onion service configuration found.
Support for v2 onion services will be removed from SecureDrop in February 2021.
Migration guide: https://securedrop.org/v2-onion-eol/
make[1]: Leaving directory '/home/kog/securedrop-workstation'
sdw-admin --apply
Applying configuration...
WARNING: v2 onion service configuration found.
Support for v2 onion services will be removed from SecureDrop in February 2021.
Migration guide: https://securedrop.org/v2-onion-eol/
gpg: can't open `/usr/share/securedrop-workstation-dom0-config/sd-journalist.sec'
Traceback (most recent call last):
  File "/usr/bin/sdw-admin", line 187, in <module>
    main()
  File "/usr/bin/sdw-admin", line 161, in main
    validate_config(SCRIPTS_PATH)
  File "/usr/bin/sdw-admin", line 89, in validate_config
    validator = SDWConfigValidator(path)  # noqa: F841
  File "/usr/share/securedrop-workstation-dom0-config/scripts/validate_config.py", line 39, in __init__
    self.confirm_submission_privkey_file()
  File "/usr/share/securedrop-workstation-dom0-config/scripts/validate_config.py", line 106, in confirm_submission_privkey_file
    subprocess.check_call(gpg_cmd, stdout=subprocess.DEVNULL)
  File "/usr/lib64/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['gpg', '/usr/share/securedrop-workstation-dom0-config/sd-journalist.sec']' returned non-zero exit status 2
Makefile:19: recipe for target 'dev' failed
make: *** [dev] Error 1
```

conorsch commented:
@zenmonkeykstop You're right, there are a bunch of `import "sd/config.json"` calls in the Salt files. The "import" path evaluates to `/srv/salt/sd/config.json`, but in other contexts such as the prod docs, we assume that `/usr/share/securedrop-workstation-dom0-config/config.json` is the proper path. We should definitely fix this, but it looks to my eye like that's a problem on main branch, too.

zenmonkeykstop commented:
`config.json` gets copied to /srv/salt/sd by the `make dev/staging` runs though, so the issue here IMO is more that `make clean` doesn't run cleanly when there isn't an existing install or the install broke before that point. If it's run before the RPM is installed then there won't be anything under  `/usr/share/securedrop...` either.

conorsch commented:
>config.json gets copied to /srv/salt/sd by the make dev/staging runs though

1. That's no longer true on this branch, which _doesn't_ copy the files to `/srv/salt/` anymore.
2. In prod contexts, we tell admins to configure the secrets in `/usr/share/`, with no mention of `/srv/salt/` at all. 

Since Salt state files require access to the json file to import, we should be using the `/usr/share/` path in all places as of this PR, since this PR uses the RPM everywhere, even for make-dev. I'll push up a commit making that change. 

> If it's run before the RPM is installed then there won't be anything under /usr/share/securedrop... either.

Unintuitive, but running 'make clean' will actually install the rpm, then remove everything, because the files for removal are installed by the RPM. That's not new in this PR, that's true even on the main branch now, where "prep-salt" is run as a dependency of make-clean.

conorsch commented:
@zenmonkeykstop The latest commit now supports make-clean even when no prior installation has been run. I did not move the import locations, preferring /usr/share/, but rather re-added the copy-to-srv-salt logic. There are two secrets, the privkey and the config.json file, and the privkey needs a "salt://` prefix (and therefore must be inside /srv/salt/`, so it wasn't worth moving just one file. The sdw-admin "install" action copies the files into place in /srv/salt on each run, and both the sdw-admin "uninstall" and the make-clean actions remove them from /srv/salt.

conorsch commented on scripts/sdw-admin.py:
```diff
@@ -113,10 +127,18 @@ def perform_uninstall():
         print("Reverting dom0 configuration")
         subprocess.check_call(["sudo", "qubesctl", "state.sls", "sd-clean-all"])
         subprocess.check_call([os.path.join(SCRIPTS_PATH, "scripts/clean-salt")])
```
Aye, that's true. Restored the `scripts/clean-salt` script since without it some of the secrets persist in `/srv/salt/`.

zenmonkeykstop replied:
:heavy_check_mark:

zenmonkeykstop reviewed (approved):
Test plan for dev and staging is passing for me now. There is one unrelated test failure in `make test` due to that race condition with systemctl in `sd-log`, but otherwise this LGTM!

conorsch commented:
> There is one unrelated test failure in make test due to that race condition with systemctl in sd-log

That'd be #583. Thanks for thorough review, @zenmonkeykstop !
